File: 11-container-with-most-water/container-with-most-water.java
class Solution {
    public int maxArea(int[] height) {
        int left = 0, right = height.length - 1;
        int result = 0;

        while (left <= right) {
            int heightOfContainer = Math.min(height[left], height[right]);
            int widthOfContainer = right - left;

            result = Math.max(result, heightOfContainer * widthOfContainer);

            while (left <= right && height[left] <= heightOfContainer)
                left++;

            while (left <= right && height[right] <= heightOfContainer)
                right--;
        }

        return result;
    }
}

File: 11-container-with-most-water/container-with-most-water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        n = len(height)
        A, l, r = 0, 0, n - 1
        while l < r:
            A = max(A, min(height[l], height[r]) * (r - l))
            if height[r] < height[l]:
                r -= 1
            else:
                l += 1
        return A
